File: modules/models/person.py
from typing import Union
from abc import ABC
from datetime import date
from modules.models.definitions import TypeOfIdentifications, Gender

class Person(ABC):
    """
    Abstract class that represents the person entity.
    """

    def __init__(self, type_of_identification: Union[TypeOfIdentifications, None], identification: Union[str, None], first_name: Union[str, None], last_name: Union[str, None], birthdate: Union[date, None], gender: Union[Gender, None], address: Union[str, None], email: Union[str, None], phone: Union[str, None]) -> None:
        """Constructor of the Person class.

        Args:
            type_of_identification (Union[TypeOfIdentifications, None]): The type of identification of the person.
            identification (Union[str, None]): The identification number of the person.
            first_name (Union[str, None]): The first name(s) of the person.
            last_name (Union[str, None]): The last name(s) of the person.
            birthdate (Union[date, None]): The birthdate of the person.
            gender (Union[Gender, None]): The gender of the person.
            address (Union[str, None]): The residential address of the person.
            email (Union[str, None]): The contact email of the person.
            phone (Union[str, None]): The telephone contact of the person.
        """
        self._type_of_identification = type_of_identification
        self._identification = identification
        self._first_name = first_name
        self._last_name = last_name
        self._birthdate = birthdate
        self._gender = gender
        self._address = address
        self._email = email
        self._phone = phone

    @property
    def type_of_identification(self) -> Union[TypeOfIdentifications, None]:
        """Get the person's ID type."""
        return self._type_of_identification
         
    @type_of_identification.setter
    def type_of_identification(self, value: Union[TypeOfIdentifications, None]) -> None:
        """Set the person's ID type."""
        self._type_of_identification = value

    @property
    def identification(self) -> Union[str, None]:
        """Get the person's identification number."""
        return self._identification
         
    @identification.setter
    def identification(self, value: Union[str, None]) -> None:
        """Set the person's identification number."""
        self._identification = value
         
    @property
    def first_name(self) -> Union[str, None]:
        """Get the person's first name."""
        return self._first_name
         
    @first_name.setter
    def first_name(self, value: Union[str, None]) -> None:
        """Set the person's first name."""
        self._first_name = value

    @property
    def last_name(self) -> Union[str, None]:
        """Get the person's last name."""
        return self._last_name
         
    @last_name.setter
    def last_name(self, value: Union[str, None]) -> None:
        """Set the person's last name."""
        self._last_name = value

    @property
    def birthdate(self) -> Union[date, None]:
        """Get the person's birthdate."""
        return self._birthdate
        
    @birthdate.setter
    def birthdate(self, value: Union[date,  None]) -> None:
        """Set the person's birthdate."""
        self._birthdate = value

    @property
    def gender(self) -> Union[Gender, None]:
        """Get the person's gender."""
        return self._gender
        
    @gender.setter
    def gender(self, value: Union[Gender, None]) -> None:
        """Set the person's gender."""
        self._gender = value

    @property
    def address(self) -> Union[str, None]:
        """Get the person's address."""
        return self._address
         
    @address.setter
    def address(self, value: Union[str, None]) -> None:
        """Set the person's address."""
        self._address = value

    @property
    def email(self) -> Union[str, None]:
        """Get the person's email."""
        return self._email
         
    @email.setter
    def email(self, value: Union[str, None]) -> None:
        """Set the person's email."""
        self._email = value

    @property
    def phone(self) -> Union[str, None]:
        """Get the person's phone."""
        return self._phone
         
    @phone.setter
    def phone(self, value: Union[str,  None]) -> None:
        """Set the person's phone."""
        self._phone = value

File: src/models/Person.ts
import { Gender, TypeOfIdentifications } from "./enums";

/**
 * @abstract Abstract class that represents a person.
 */
export default abstract class Person {
  /**
   * Constructor of the abstract class Person.
   *
   * @param {TypeOfIdentification | undefined} _typeOfIdentification - Type of identification of the person.
   * @param {string | undefined} _identification - Identification number of the person.
   * @param {string | undefined} _firstName - First name of the person.
   * @param {string | undefined} _lastName - Last name of the person.
   * @param {Date | undefined} _birthdate - Date of birth of the person.
   * @param {Gender | undefined} _gender - Gender of the person.
   * @param {string | undefined} _address - Address of the person.
   * @param {string | undefined} _email - Email of the person.
   * @param {string | undefined} _phone - Person's telephone number.
   */
  constructor(
    protected _typeOfIdentification?: TypeOfIdentifications,
    protected _identification?: string,
    protected _firstName?: string,
    protected _lastName?: string,
    protected _birthdate?: Date,
    protected _gender?: Gender,
    protected _address?: string,
    protected _email?: string,
    protected _phone?: string
  ) {}

  /**
   * Gets the type of identification of the person.
   *
   * @returns {TypeOfIdentifications | undefined} The type of identification of the person
   */
  get typeOfIdentification(): TypeOfIdentifications | undefined {
    return this._typeOfIdentification;
  }

  /**
   * Sets the type of identification of the person.
   *
   * @param {TypeOfIdentifications | undefined} value - The type of identification of the person.
   */
  set typeOfIdentification(value: TypeOfIdentifications | undefined) {
    this._typeOfIdentification = value;
  }

  /**
   * Gets the identification number of the person.
   *
   * @returns {string | undefined} The identification number of the person.
   */
  get identification(): string | undefined {
    return this._identification;
  }

  /**
   * Sets the identification number of the person.
   *
   * @param {string | undefined} value - The identification number of the person.
   */
  set identification(value: string | undefined) {
    this._identification = value;
  }

  /**
   * Gets the first name of the person.
   *
   * @returns {string | undefined} The first name of the person.
   */
  get firstName(): string | undefined {
    return this._firstName;
  }

  /**
   * Sets the first name of the person.
   *
   * @param {string | undefined} value - The first name of the person.
   */
  set firstName(value: string | undefined) {
    this._firstName = value;
  }

  /**
   * Gets the last name of the person.
   *
   * @returns {string | undefined} The last name of the person.
   */
  get lastName(): string | undefined {
    return this._lastName;
  }

  /**
   * Sets the last name of the person.
   *
   * @param {string | undefined} value - The last name of the person.
   */
  set lastName(value: string | undefined) {
    this._lastName = value;
  }

  /**
   * Gets the date of birth of the person.
   *
   * @returns {Date | undefined} The date of birth of the person.
   */
  get birthdate(): Date | undefined {
    return this._birthdate;
  }

  /**
   * Sets the date of birth of the person.
   *
   * @param {Date | undefined} value - The date of birth of the person.
   */
  set birthdate(value: Date | undefined) {
    this._birthdate = value;
  }

  /**
   * Gets the gender of the person.
   *
   * @returns {Gender | undefined} The gender of the person.
   */
  get gender(): Gender | undefined {
    return this._gender;
  }

  /**
   * Sets the gender of the person.
   *
   * @param {Gender | undefined} value - The gender of the person.
   */
  set gender(value: Gender | undefined) {
    this._gender = value;
  }

  /**
   * Gets the address of the person.
   *
   * @returns {string | undefined} The address of the person.
   */
  get address(): string | undefined {
    return this._address;
  }

  /**
   * Sets the address of the person.
   *
   * @param {string | undefined} value - The address of the person.
   */
  set address(value: string | undefined) {
    this._address = value;
  }

  /**
   * Gets the email of the person.
   *
   * @returns {string | undefined} The email of the person.
   */
  get email(): string | undefined {
    return this._email;
  }

  /**
   * Sets the email of the person.
   *
   * @param {string | undefined} value - The email of the person.
   */
  set email(value: string | undefined) {
    this._email = value;
  }

  /**
   * Gets the phone number of the person.
   *
   * @returns {string | undefined} The phone number of the person.
   */
  get phone(): string | undefined {
    return this._phone;
  }

  /**
   * Sets the phone number of the person.
   *
   * @param {string | undefined} value - The phone number of the person.
   */
  set phone(value: string | undefined) {
    this._phone = value;
  }

  /**
   * @abstract Abstract method that must be implemented by child classes to validate the person's data.
   *
   * @returns {boolean} True if the person's data is valid, false otherwise.
   */
  abstract validate(): boolean;
}
